sql/core/src/main/scala/org/apache/spark/sql/execution/RecursiveCTEExecution.scala

```diff
@@ -0,0 +1,229 @@
+                          localLimit: Option[Int] = None,
+                          globalLimit: Option[Int] = None) extends LeafExecNode {
+
+  override def innerChildren: Seq[QueryPlan[_]] = Seq(anchor, recursion)
+
+  private val numPartitions: Int = conf.defaultNumShufflePartitions
+
+  override lazy val metrics = Map(
+    "numOutputRows" -> SQLMetrics.createMetric(sparkContext, "number of output rows"),
+    "numRecursiveLoops" -> SQLMetrics.createMetric(sparkContext, "number of recursive loops"))
+
+  private val simpleRecursion = {
+    recursion match {
+      case Project(_, Filter(_, Project(_, UnionLoopRef(_, _, _)))) =>
+        true
+      case Filter(_, Project(_, UnionLoopRef(_, _, _))) =>
+        true
+      case Project(_, Filter(_, UnionLoopRef(_, _, _))) =>
+        true
+      case Project(_, UnionLoopRef(_, _, _)) =>
+        true
+      case _ =>
+        false
+  }
+  }
+  /**
+   * This function executes the plan (optionally with appended limit node) and caches the result,
+   * with the caching mode specified in config.
+   */
+  private def executeAndCacheAndCount(
+                                       plan: LogicalPlan, currentLimit: Int) = {
+    // In case limit is defined, we create a (local) limit node above the plan and execute
+    // the newly created plan.
+    val planOrLimitedPlan = if (globalLimit.isDefined || localLimit.isDefined) {
+      LocalLimit(Literal(currentLimit), plan)
+    } else {
+      plan
+    }
+    val df = Dataset.ofRows(session, planOrLimitedPlan)
+    val newDF = {
+      if (!simpleRecursion) {
+        df.repartition(numPartitions)
+      } else {
+        df
+      }
+    }
+    val count = newDF.count()
+    (newDF, count)
+  }
+
+  /**
+   * In the first iteration, anchor term is executed.
+   * Then, in each following iteration, the UnionLoopRef node is substituted with the plan from the
+   * previous iteration, and such plan is executed.
+   * After every iteration, the dataframe is repartitioned.
+   * The recursion stops when the generated dataframe is empty, or either the limit or
+   * the specified maximum depth from the config is reached.
+   */
+  override protected def doExecute(): RDD[InternalRow] = {
+    val executionId = sparkContext.getLocalProperty(SQLExecution.EXECUTION_ID_KEY)
+    val numOutputRows = longMetric("numOutputRows")
+    val numRecursiveLoops = longMetric("numRecursiveLoops")
+    val levelLimit = conf.getConf(SQLConf.CTE_RECURSION_LEVEL_LIMIT)
+
+    // currentLimit is initialized from the limit argument, and in each step it is decreased by
+    // the number of rows generated in that step.
+    // If limit is not passed down, currentLimit is set to be zero and won't be considered in the
+    // condition of while loop down (limit.isEmpty will be true).
+    var globalLimitNum = globalLimit.getOrElse(0)
+    var localLimitNum = localLimit.getOrElse(0)
+    var currentLimit = Math.max(globalLimitNum, localLimitNum * numPartitions)
```

Comment:
I don't think `localLimitNum * numPartitions` can be treated as a global limit. Think about this case:
1. The query result has 2 partitions. In each iteration, the first partition produces 1 million rows, and the second partition produces 1 row.
2. Let's say the local limit is 100. We need to iterate 100 times so that the second partition produces enough data, but the current code stops at the first iteration because 1 million already exceeds 100 * 2.